src/lib/zf/attr.c

```diff
@@ -126,6 +126,7 @@ static struct zf_attr* __zf_attr_alloc(void)
 int zf_attr_alloc(struct zf_attr** attr_out)
 {
   struct zf_attr* attr = __zf_attr_alloc();
+  zf_log_stderr(); /*log_file is static and we want default behaviour for new set of attributes*/
```

Comment:
If an app creates different sets of attributes for different stacks, upon creating a second set of attributes it is going to change logging for the first since zf_log_file is global.

Not sure this is the behaviour we want.

Same applies to zf_log_level and zf_log_format.

Also note that zf_attr_dup() does not pass through here so it gives a path to create a new set of attributes with different behaviour to creating two sets manually.

I'm thinking a cleaner approach might be to replace global zf_log_level, etc with a global set of attributes that are only initialised in zf_init() and a call to modify any global attributes modifies only the global ones.  I don't think changing "local" attributes to then modify global values makes sense.  All the local ones could be different values with only one sets of attributes matching the global ones.

Sorry for not picking this up earlier but only thought through it when seeing this call here now.

Reply:
zf_log_file, zf_log_level and zf_log_format were initially intended to be global before ON-15400 and the current behaviour is just inherited. If this is not the case, I think we should raise a separate case for this, or are you proposing to forbid the zf_attr_set_str( ) for log_file, and to forbid zf_attr_set_int( ) for log_level and log_format to be directly called from the user's code?
I'm a bit confused as the proposed code in this PR (ON-15400) do not change "local" attributes, it changes the global attributes as originally intended, but basically the code doing that change was just missing and, now, this PR adds it.

Reply:
Currently (prior to any of your changes) all calls to zf_attr_set_...(attr) operate only on attributes within attr and different stacks can use different attributes.  The log_file, log_format, log_level are global attributes and changing them will modify these settings across all stacks not only for a stack related to the attr it is called on.  I think this is ok if well explained / documented.

What I do think we need to change is that when zf_attr_alloc() is called, logging changes will get reset and I think this would be unexpected.  I.e. this sequence

1. zf_attr_alloc(&attr1)
2. zf_attr_set_...(attr1, <change logging settings>)
3. zf_stack_alloc(attr1, &stack1)
4. zf_attr_alloc(&attr2) <=== at this point I expect the global logging options already configured to get reset with the current implementation.  Instead I think global attributes should only be initialised when calling zf_init() or when explicitly modified with a zf_attr_set_...() call.

At the least we should have tests for above sequence to check the behaviour and whether my interpretation is correct.

Also consider the sequence when in step 4. you create attr2 by calling zf_attr_dup() on attr1.  This follows a different code path and would currently give a different behaviour to two calls to zf_attr_alloc() with regards to resetting logging options which I don't think should be the case.

Reply:
Yes, agreed. Missing a bit of documentation for the new behaviour when modifying log_file, log_format and log_level as these are global attributes.
Also, it is true that allocating a new set of attributes (via zf_attr_alloc) will reset the global log_file to stderr and the proposed sequence of calls end up in an unintended behaviour. The problem is that the program line 129 referenced above should not be there. Removing it fixes the problem.

Also, checked zf_attr_dup and it does respect the new behaviour, ie, although it does initialize the log_file to stderr at allocation time, and copy the log_file value from the source "attr", it does not modifies the zf_log_file  static value. Hence, the existent value in zf_log_file remains intact after calling zf_attr_dup. Good test to consider, though. 

```
struct zf_attr* zf_attr_dup(const struct zf_attr* from)
{
  struct zf_attr* to = __zf_attr_alloc();
  if( to != NULL )
    zf_attr_copy(to, from);
  return to;
}  
```